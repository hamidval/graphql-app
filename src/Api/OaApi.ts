

export class OaApi 
{
    constructor(connectionString: string)
    {

    

    }
}