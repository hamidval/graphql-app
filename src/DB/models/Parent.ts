import * as Sequelize from "sequelize";
import { DataTypes, Model, Optional } from 'sequelize';
import { Student } from "./Student";

export class Parent extends Model
{
    students:Student[]
    Id:number;
    FirstName: String;
    LastName: String;
    Email:String;
    static initModel(sequelize: Sequelize.Sequelize): typeof Parent 
    {
        
        return Parent.init(
            {
                Id: {
                    autoIncrement: true,
                    type: DataTypes.INTEGER,
                    allowNull: false,
                    primaryKey: true
                },
                FirstName: {
                    autoIncrement: false,
                    type: DataTypes.TEXT,
                    allowNull: false,
                    primaryKey: true
                },
                LastName: {
                    autoIncrement: false,
                    type: DataTypes.TEXT,
                    allowNull: false,
                    primaryKey: true
                },
                Email: {
                    autoIncrement: false,
                    type: DataTypes.TEXT,
                    allowNull: false,
                    primaryKey: true
                }
            },
            {
                sequelize,
                tableName: 'Parent',
                schema: 'dbo',
                timestamps: false,
                indexes: []
              });
    }
}