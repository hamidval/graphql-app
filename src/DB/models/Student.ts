import * as Sequelize from "sequelize";
import { DataTypes, Model, Optional } from 'sequelize';
import { Parent } from "./Parent";
import { Lesson } from "./Lesson";
import { TakenLesson } from "./TakenLesson";

export class Student extends Model
{
    parent:Parent;
    Lessons:Lesson[];
    TakenLessons: TakenLesson[];
    Id: number;
    FirstName: String;
    LastName: String;
    static initModel(sequelize: Sequelize.Sequelize): typeof Student 
    {
        
        return Student.init(
            {
                Id: {
                    autoIncrement: true,
                    type: DataTypes.INTEGER,
                    allowNull: false,
                    primaryKey: true
                },
                FirstName: {
                    autoIncrement: false,
                    type: DataTypes.TEXT,
                    allowNull: false,
                    primaryKey: true
                },
                LastName: {
                    autoIncrement: false,
                    type: DataTypes.TEXT,
                    allowNull: false,
                    primaryKey: true
                }
            },
            {
                sequelize,
                tableName: 'Student',
                schema: 'dbo',
                timestamps: false,
                indexes: []
              });
    }
}