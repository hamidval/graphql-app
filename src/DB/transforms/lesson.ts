import { Lesson } from "../models/Lesson";
import {Lesson as _Lesson} from "../../generated/graphql"

export default class LessonTransforms 
{
    constructor()
    {

    }

    public lessons(lessons: Lesson[]) : _Lesson[]
    {
   
        return lessons.map((x)=> this.lesson(x));
       

    }

    private lesson(lesson: Lesson): _Lesson
    {
        return  {
            Id: lesson.Id,
            Day: lesson.Day,
            Subject: lesson.Subject,
            IsALevel:this.isALevel(lesson.Subject),
            IsWeekend: this.isWeekend(lesson.Day),
            SingleFee: lesson.SingleFee,
            SinglePay: lesson.SinglePay,
            GroupFee: lesson.GroupFee,
            GroupPay: lesson.GroupPay
        };
    }

    private isALevel(subject: number): boolean
    {
        if(subject == 0 || subject == 1 || subject == 2)
        {
            return true;
        }

        return false;
    }

    private isWeekend(day: number): boolean
    {
        if(day == 6 || day == 0)
        {
            return true;
        }

        return false;
    }
}