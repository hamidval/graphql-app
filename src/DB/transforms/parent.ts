

import { Parent } from "../models/Parent";
import { Parent as _Parent } from "../../generated/graphql"

export default class ParentTransforms
{
    public parents(parents: Parent[]) : _Parent[]
    {
        return parents.map((x) => this.parent(x));
    }

    public parent(parent: Parent): _Parent
    {
        return {
            Id: parent.Id,
            FirstName: parent.FirstName,
            LastName: parent.LastName,
            Email: this.email(parent.Email)

        }
    }

    private email(email: String): String
    {
        return email.slice(0,6) + "*******"
    }
}