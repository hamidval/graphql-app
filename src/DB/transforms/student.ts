import { Student } from "../models/Student";
import {Student as _Student}  from "../../generated/graphql"

export default class StudentTransforms 
{
    constructor(){}
    
    
    public students(students: Student[]):_Student[] {

        return students.map(x => this.student(x))
    }


    public student(student: Student): _Student{
        return{
            Id: student.Id,
            FirstName: student.FirstName,
            LastName: student.LastName,
            ParentId: student.parent.Id
        }
    }   
}