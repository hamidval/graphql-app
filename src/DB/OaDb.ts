import { FindOptions } from "sequelize";
import { SequelizeDataSource } from "./index";
import { Lesson } from "./models/Lesson";
import { Student } from "./models/Student";
import { Parent } from "./models/Parent";
import { TakenLesson } from "./models/TakenLesson"
import { Where } from "sequelize/types/utils";
import { whiteList } from "../utils";
import { Teacher } from "./models/Teacher";
import { TeacherTransform } from "./transforms/teacher";
import TakenLessonTransforms from "./transforms/takenLesson";


type where ={
  id?: number | null | number[],
  day?: number,
  studentId?: number,
  teacherId?: number | number[],
  parentId?:number | number [] 
}

export default class OaDb extends SequelizeDataSource 
{
    public async getParents(id:number | null = null)
    {
      var query:FindOptions = {attributes:['Id', 'FirstName', 'LastName', 'Email']}

      if(id == null)
      {
        query.limit = 10;
      }else
      {
        query.where = {id:id}       
      }  

      var data = await Parent.findAll(query)
      
      return data;
    
    }

      public async getStudents(parentId:any): Promise<any[]> {    
        var data =  (await this.client.query(`SELECT * FROM Student WHERE ParentId = ${parentId}`, {
            type: 'SELECT'
          })) as Array<any>;   

       return data;
     }


     
     public async getAllStudents(id:number | null = null, parentId:number | null = null,  inlcudeLessons: boolean = false)
     {
      var query:FindOptions = {}
      let where: where = {}

      if(id == null && parentId == null)
      {
        query.limit = 10;
      }
      
      if(id) {
        where.id = id       
      }

      if(parentId){
        where.parentId = parentId
      }

      if(inlcudeLessons)
      {
        query.include = [{model:Lesson, as: "Lessons"}]
      }     

      query.where = where;
      var data = await Student.findAll(query)
      return data;
      
     }

     public async getLessons(id:number | null = null, day: number | null, studentId: number | null, teacherId: number | null)
     {
      var query:FindOptions = {}

      let _where:where = {}
      _where.teacherId = whiteList();
      if(id == null && studentId == null && day == null && teacherId == null)
      {
        query.limit = 10;
      }else
      {
        
        if(id) {
          _where.id = id  
        }

        if(studentId) {
          _where.studentId = studentId
        }

        if(day){
          _where.day = day
        }

        if(teacherId)
        {
          _where.teacherId = whiteList(teacherId);
        }
             
      }  
      query.where = _where;
      var data = await Lesson.findAll(query)
      
      return data;
      
     }

     public async getTakenLessons(
      id:number | null = null, 
      teacherId: number | null = null, 
      studentId: number | null = null)
     {
      var query:FindOptions = {}

      let _where:where = {}
      _where.teacherId = whiteList();
      if(id == null && teacherId == null && studentId == null)
      {
        query.limit = 10;
      }
      if(id) {
        _where = {id:id}       
      }
      
      if(teacherId){
        _where.teacherId = whiteList(teacherId);
      }

      if(studentId){
        _where.studentId = studentId
      }

      query.where = _where;
      var data = new TakenLessonTransforms().takenLessons(await TakenLesson.findAll(query));
      
      return data;
      
     }

     public async getAllTeachers(id:number | null)
     {
        let query:FindOptions = {}
        let _where:where = {}
        
        if(id != null){
          _where.id = whiteList(id);
        }
        else 
        {
          _where.id = whiteList()
        }

        query.where = _where;
        var data = await Teacher.findAll(query);
        return new TeacherTransform().teachers(data)
     }

     public async updateLessonSubject(id:number, subject:number) : Promise<Lesson>
     {
        var lesson  =  await Lesson.findOne({where:{id:id}});
        lesson.update({Subject: subject})
        lesson.save();
        return lesson;
     }

      public async connect(): Promise<string> {
        try {
          await this.client.authenticate();
    
          return `Connected to`;
        } catch (error) {
          return error;
        }
      }
}