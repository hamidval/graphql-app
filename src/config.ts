import OaDb from "./DB/OaDb";
import { OaApi } from "./queries/queries";
import { sequelizeOa } from './DB/singlton';
import { resolvers } from "./schema";
import { typeDef as Lesson } from "./schema/types/lesson";
import { typeDef as Parent } from "./schema/types/parent";
import { typeDef as Student } from "./schema/types/student";
import { typeDef as LessonInput } from "./schema/types/lessonInput";
import { typeDef as StudentInput } from "./schema/types/studentInput";
import { typeDef as ParentInput } from "./schema/types/parentInput";
import { typeDef as QueryType } from "./schema/types/queryType";
import { typeDef as UpdateLessonSubjectType } from "./schema/types/mutations/updateLessonSubject";
import { typeDef as UpdateLessonSubjectInputType } from "./schema/types/mutations/updateLessonSubjectInput";
import { typeDef as TakenLesson } from "./schema/types/takenLesson";
import { typeDef as Teacher } from "./schema/types/teacher";

export interface RequestHeaders {
    'accept-language': string;
  }

export class RequestQuery {
    body: string;
  
    constructor(body?: string) {
      this.body = body ?? '';
    }
  
    contains = function (node: string) {
      return this.body.match(new RegExp(`\\b${node}\\b`, 'i'));
    };
  }
  
  export interface Request {
    headers: RequestHeaders;
    query: RequestQuery;
  }
  

export interface Context  {
    dataSources: {
        OaApi: OaApi;
        OaDb: OaDb;
    };
    request: Request;
  }
 
const dataSources = (): Context["dataSources"] => {
    return {
      OaApi: new OaApi(),
      OaDb: (new OaDb(sequelizeOa))
  };
}

export const config:any = 
{ 
    typeDefs:[
        UpdateLessonSubjectInputType, UpdateLessonSubjectType, Parent, Student, Lesson,
        LessonInput, StudentInput, ParentInput,
        QueryType, TakenLesson, Teacher
        ],
    dataSources,
    resolvers
}