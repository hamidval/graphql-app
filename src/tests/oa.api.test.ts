

import { resolvers } from '../schema'
const { ApolloServer } = require('apollo-server');
import { typeDef as Lesson } from "../schema/types/lesson";
import { typeDef as Parent } from "../schema/types/parent";
import { typeDef as Student } from "../schema/types/student";
import { typeDef as LessonInput } from "../schema/types/lessonInput";
import { typeDef as StudentInput } from "../schema/types/studentInput";
import { typeDef as ParentInput } from "../schema/types/parentInput";
import { typeDef as QueryType } from "../schema/types/queryType";
import { typeDef as UpdateLessonSubjectType } from "../schema/types/mutations/updateLessonSubject";
import { typeDef as UpdateLessonSubjectInputType } from "../schema/types/mutations/updateLessonSubjectInput";
import { typeDef as TakenLesson } from "../schema/types/takenLesson";
import { typeDef as Teacher } from "../schema/types/teacher";

import {Parent as ParentModel } from '../DB/models/Parent'
import {Student as StudentModel } from '../DB/models/Student'

import { QUERY_STUDENT, QUERY_PARENT, QUERY_LESSON, QUERY_TAKEN_LESSON } from './queries';
import OaDb from "../DB/OaDb";
import { Sequelize } from "sequelize";
import { Context } from "../config";
import { OaApi } from "../Api/OaApi";
const  testServer = (dataSources) : typeof ApolloServer =>
{
    return new ApolloServer({
        typeDefs:[UpdateLessonSubjectInputType, UpdateLessonSubjectType, Parent, Student, Lesson,
            LessonInput, StudentInput, ParentInput,
            QueryType, TakenLesson, Teacher],
        resolvers,
        dataSources
        });
}




describe('resolvers', () =>
{
    let oaDb;
    let datasources;
    let server;

    beforeEach(()=>
    {
        let seq = new Sequelize('', '', '', { host: '', port: 0, database: '', username: '', password: '', dialect: 'mssql' })
        oaDb = new OaDb(seq);

        datasources = (): Context["dataSources"] => {
            return {
              OaApi: new OaApi(''),
              OaDb: oaDb
          };
        }

        server = testServer(datasources);
    })

    it('test parents', async () =>
    {      

        oaDb.getParents = jest.fn(()=>[]);

        const res = await server.executeOperation({
            query: QUERY_PARENT,
            variables: { } as object
          });      
          expect(res.errors).toBe(undefined);      
         // expect(res).toMatchSnapshot();
    }); 

    it('test students', async () =>
    {

        oaDb.getParents = jest.fn(()=>[]);

        //spys on this Model
        jest.spyOn(oaDb, 'getAllStudents').mockImplementation(() =>{
            return Promise.all([]);
        })

        const res = await server.executeOperation({
            query: QUERY_STUDENT,
            variables: { "studentInput": { "filter":{"id": 1, "includeLessons": null, "parentId": null}}} as object
          });      
          expect(res.errors).toBe(undefined);      
          //expect(res).toMatchSnapshot();
    });


    it('test lessons', async () =>
    {
        oaDb.getParents = jest.fn(()=>[]);

        //spys on this Model
        jest.spyOn(oaDb, 'getLessons').mockImplementation(() =>{
            return Promise.all([]);
        })

        const res = await server.executeOperation({
            query: QUERY_LESSON,
            variables: { "lessonInput": { "filter":{"teacherId": 1}}} as object
          });      
          expect(res.errors).toBe(undefined);      
          //expect(res).toMatchSnapshot();
    }); 


    it('test taken lessons', async () =>
    {
        oaDb.getParents = jest.fn(()=>[]);

        //spys on this Model
        jest.spyOn(oaDb, 'getTakenLessons').mockImplementation(() =>{
            return Promise.all([]);
        })

        const res = await server.executeOperation({
            query: QUERY_TAKEN_LESSON,
            variables: { "takenLessonInput": { "filter":{"id": null}}} as object
          });      
          expect(res.errors).toBe(undefined);      
         // expect(res).toMatchSnapshot();
    }); 

    it('get lessons by teacher id', async () =>
    {
        //spys on this Model
        jest.spyOn(oaDb, 'getLessons').mockImplementation(() =>{
            return Promise.all([{"Id": 526},{"Id": 979},{"Id": 1010 },{"Id": 1023 }, { "Id": 1030},
              {"Id": 1038 }, { "Id": 1089 }, {
                "Id": 1122
              },
              {
                "Id": 1138
              },
              {
                "Id": 1174
              },
              {
                "Id": 1178
              },
              {
                "Id": 1248
              },
              {
                "Id": 1320
              },
              {
                "Id": 1335
              },
              {
                "Id": 1337
              },
              {
                "Id": 1344
              },
              {
                "Id": 1359
              },
              {
                "Id": 1409
              },
              {
                "Id": 1410
              },
              {
                "Id": 1420
              },
              {
                "Id": 1421
              },
              {
                "Id": 1508
              },
              {
                "Id": 1569
              },
              {
                "Id": 1575
              },
              {
                "Id": 1584
              },
              {
                "Id": 1625
              },
              {
                "Id": 1640
              }]);
        })

        const res = await server.executeOperation({
            query: QUERY_LESSON,
            variables: { "lessonInput": { "filter":{"teacherId": 1}}} as object
          });      
          expect(res.errors).toBe(undefined);   
          expect(res.data.lessons.length).toBe(27);      
          //expect(res).toMatchSnapshot();
    }); 
});