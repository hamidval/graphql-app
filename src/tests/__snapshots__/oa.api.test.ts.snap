// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolvers get lessons by teacher id 1`] = `
{
  "data": {
    "lessons": [
      {
        "Id": 526,
      },
      {
        "Id": 979,
      },
      {
        "Id": 1010,
      },
      {
        "Id": 1023,
      },
      {
        "Id": 1030,
      },
      {
        "Id": 1038,
      },
      {
        "Id": 1089,
      },
      {
        "Id": 1122,
      },
      {
        "Id": 1138,
      },
      {
        "Id": 1174,
      },
      {
        "Id": 1178,
      },
      {
        "Id": 1248,
      },
      {
        "Id": 1320,
      },
      {
        "Id": 1335,
      },
      {
        "Id": 1337,
      },
      {
        "Id": 1344,
      },
      {
        "Id": 1359,
      },
      {
        "Id": 1409,
      },
      {
        "Id": 1410,
      },
      {
        "Id": 1420,
      },
      {
        "Id": 1421,
      },
      {
        "Id": 1508,
      },
      {
        "Id": 1569,
      },
      {
        "Id": 1575,
      },
      {
        "Id": 1584,
      },
      {
        "Id": 1625,
      },
      {
        "Id": 1640,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`resolvers random test 1`] = `
{
  "data": {
    "students": [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`resolvers test lessons 1`] = `
{
  "data": {
    "lessons": [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`resolvers test parents 1`] = `
{
  "data": {
    "parents": [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`resolvers test students 1`] = `
{
  "data": {
    "students": [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`resolvers test taken lessons 1`] = `
{
  "data": {
    "takenLessons": [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
